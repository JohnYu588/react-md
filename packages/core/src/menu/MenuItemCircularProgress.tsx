import { cnb } from "cnbuilder";
import { forwardRef, type HTMLAttributes, type ReactNode } from "react";
import { ListItemChildren } from "../list/ListItemChildren.js";
import { listItem } from "../list/listItemStyles.js";
import {
  type ListItemChildrenProps,
  type ListItemHeight,
} from "../list/types.js";
import {
  CircularProgress,
  type CircularProgressProps,
} from "../progress/CircularProgress.js";

/**
 * @since 6.0.0
 */
export interface MenuItemCircularProgressProps extends CircularProgressProps {
  /** @defaultValue `"Loading"` */
  "aria-label"?: string;

  /** @defaultValue `"none"` */
  role?: string;

  /** @defaultValue `true` */
  dense?: boolean;

  liProps?: HTMLAttributes<HTMLLIElement>;
  height?: ListItemHeight;
  leftAddon?: ReactNode;
  rightAddon?: ReactNode;
  listItemChildrenProps?: ListItemChildrenProps;
}

/**
 * @example Simple Example
 * ```tsx
 * "use client";
 * import {
 *   DropdownMenu,
 *   MenuItem,
 *   MenuItemCircularProgress,
 *   useAsyncFunction,
 * } from "@react-md/core";
 *
 * function Example() {
 *   const { handleAsync, pending } = useAsyncFunction();
 *   const [visible, setVisible] = useState(false);
 *
 *   return (
 *     <DropdownMenu buttonChildren="Options" visible={visible} setVisible={setVisible}>
 *       {pending && <MenuItemCircularProgress />}
 *       <MenuItem
 *         onClick={handleAsync(async (event) => {
 *           // prevent menu from closing until action completes
 *           event.stopPropagation()
 *
 *           // do something async
 *           await ...
 *
 *           // close menu once completed
 *           setVisible(false)
 *         })}
 *       >
 *         Some Action
 *       </MenuItem>
 *     </DropdownMenu>
 *   )
 * }
 * ```
 *
 * @since 6.0.0
 */
export const MenuItemCircularProgress = forwardRef<
  HTMLLIElement,
  MenuItemCircularProgressProps
>(function MenuItemCircularProgress(props, ref) {
  const {
    "aria-label": propAriaLabel,
    "aria-labelledby": ariaLabelledBy,
    role = "none",
    className,
    height,
    leftAddon,
    rightAddon,
    liProps,
    listItemChildrenProps,
    ...remaining
  } = props;
  let ariaLabel = propAriaLabel;
  if (!ariaLabel && !ariaLabelledBy) {
    ariaLabel = "Loading";
  }

  return (
    <li
      {...liProps}
      ref={ref}
      role={role}
      className={cnb(listItem({ height, className }), liProps?.className)}
    >
      <ListItemChildren
        {...listItemChildrenProps}
        leftAddon={leftAddon}
        rightAddon={rightAddon}
      >
        <CircularProgress
          aria-label={ariaLabel as string}
          aria-labelledby={ariaLabelledBy}
          dense
          {...remaining}
        />
      </ListItemChildren>
    </li>
  );
});
