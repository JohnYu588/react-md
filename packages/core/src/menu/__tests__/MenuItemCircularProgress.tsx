import { describe, expect, it } from "@jest/globals";
import { type ReactElement, createRef } from "react";

import { rmdRender, screen } from "../../test-utils/index.js";
import { MenuItemCircularProgress } from "../MenuItemCircularProgress.js";

function render(ui: ReactElement) {
  return rmdRender(ui, {
    wrapper: function Wrapper({ children }) {
      return <ul>{children}</ul>;
    },
  });
}

describe("MenuItemCircularProgress", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLLIElement>();
    const props = {
      liProps: {
        "data-testid": "li",
      },
      ref,
    } as const;
    const { rerender } = render(<MenuItemCircularProgress {...props} />);

    const li = screen.getByTestId("li");
    expect(ref.current).toBeInstanceOf(HTMLLIElement);
    expect(ref.current).toBe(li);
    expect(li).toMatchSnapshot();

    rerender(
      <MenuItemCircularProgress
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(li).toMatchSnapshot();
  });
});
