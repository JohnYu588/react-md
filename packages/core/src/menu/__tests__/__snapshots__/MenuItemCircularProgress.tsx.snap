// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemCircularProgress should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<li
  class="rmd-list-item rmd-interaction-surface"
  data-testid="li"
  role="none"
>
  <span
    class="rmd-list-item__text"
  >
    <span
      aria-label="Loading"
      aria-valuemax="100"
      aria-valuemin="0"
      class="rmd-circular-progress rmd-circular-progress--dense rmd-circular-progress--centered rmd-primary-color"
      id="circular-progress-:r0:"
      role="progressbar"
    >
      <svg
        class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
        viewBox="0 0 66 66"
      >
        <circle
          class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
          cx="33"
          cy="33"
          r="30"
        />
      </svg>
    </span>
  </span>
</li>
`;

exports[`MenuItemCircularProgress should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<li
  class="rmd-list-item rmd-interaction-surface custom-class-name"
  data-testid="li"
  role="none"
>
  <span
    class="rmd-list-item__text"
  >
    <span
      aria-label="Loading"
      aria-valuemax="100"
      aria-valuemin="0"
      class="rmd-circular-progress rmd-circular-progress--dense rmd-circular-progress--centered rmd-primary-color"
      id="circular-progress-:r0:"
      role="progressbar"
      style="color: white;"
    >
      <svg
        class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
        viewBox="0 0 66 66"
      >
        <circle
          class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
          cx="33"
          cy="33"
          r="30"
        />
      </svg>
    </span>
  </span>
</li>
`;
